WordPress/gutenberg
# Dimensions: Add Aspect Ratio block support

## What?


Fixes: https://github.com/WordPress/gutenberg/issues/54032

This PR explores adding aspect ratio to the Cover and Group blocks by including it as part of the dimensions block supports.

It supports aspect ratio at the individual block level, and within global styles. Note that the override for the Cover block that switches off its min-height value only works at the individual block level, and not in global styles.



## Why?


We currently only have minHeight controls, which doesn't allow for easily creating particular proportions. It'd be helpful for many different kinds of site designs to be able to set container blocks like Cover and Group to a particular aspect ratio.

## How it works (and potential limitations)

This PR tries to let the CSS `aspect-ratio` work the way that it works, with minimal interference. To achieve this, and to avoid overflow issues, when an `aspect-ratio` is set, we `unset` any value for `min-height`. Similarly, if a `min-height` is set at the block level, we output an `aspect-ratio: unset` rule to override any aspect ratio set in global styles. These `unset` rules should only be output for blocks that opt-in to aspect ratio, and are not saved to post content.

For folks wanting to reduce the width of their container to less than the full width of the container, a good way to do it is to wrap the block in a Row block, and use the child width controls, as the flex-basis rules for fixed width appear to play nicely here.

## How?


* Add `aspectRatio` to dimensions block support
* Opt the Cover block into that support (note: Group block support was removed in #58414)
* Add a `min-height: unset` rule that gets applied when aspect ratio is in use — this is necessary because setting a `min-height` interferes with aspect-ratio behaviour by not allowing content to expand the area of a block beyond the defined aspect ratio. In order to ensure that longer content flows correctly in narrower viewports, we unset `min-height` when aspect ratio it applied.
* This PR also uses server-rendering instead of saving the `aspect-ratio` rule into post content. This is because the aspect ratio rule depends on other CSS values, so it's safer if we don't serialize the value for now. This also allows us to explore other options in the future for output, like using `has-4-3-aspect-ratio` classnames that apply the aspect ratio instead of inline styles. And finally, it allows us to add `aspect-ratio` and `unset` rules without having to deprecate any legacy markup.
* Update global styles output to _also_ output a `min-height: unset` rule when outputting `aspect-ratio` — this is a little hacky, but ensures that rules like the Cover block's default `min-height` don't interfere with the aspect ratio.
* Note: for now, the Group block's aspect ratio controls are disabled in global styles, to prevent users from unexpectedly setting aspect ratio for all of the group block's variations



## Testing Instructions





1. Add Cover and Group blocks to a post or page.
2. Set background images and colors on the blocks.
3. Try selecting from a range of aspect ratios under the Dimensions panel.
4. Try playing around with adding `minHeight`, too — notice that `minHeight` will clear out any aspect ratio applied.
5. To reduce the overall size of the block, try wrapping in a Row block, and use the Fixed width controls to reduce the size of the block.
6. Also, try playing around with aspect ratio at the block level in global styles.



In testing, how do these controls feel to use? Is it close enough to desired behaviour to be useful, or does it feel limiting? Try adding a variety of content within the Group and Cover blocks to get a feel for it, too.

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/14988353/26d749c1-6bd5-403a-b844-c323db8cf644
